src/kerberos_sspi.c

```diff
@@ -358,6 +457,126 @@ auth_sspi_client_step(sspi_client_state* state, SEC_CHAR* challenge, SecPkgConte
     return status;
 }
 
+INT
+auth_sspi_server_step(sspi_server_state* state, SEC_CHAR* challenge, SecPkgContext_Bindings* sec_pkg_context_bindings) {
+	SecBufferDesc inbuf;
+	SecBuffer inBufs[2];
+	SecBufferDesc outbuf;
+	SecBuffer outBufs[1];
+	ULONG ignored;
+	SECURITY_STATUS status = AUTH_GSS_CONTINUE;
+	DWORD len;
+	BOOL haveToken = FALSE;
+	INT tokenBufferIndex = 0;
+
+	if (state->response != NULL) {
+		free(state->response);
+		state->response = NULL;
+	}
+
+	inbuf.ulVersion = SECBUFFER_VERSION;
+	inbuf.pBuffers = inBufs;
+	inbuf.cBuffers = 0;
+
+	if (sec_pkg_context_bindings != NULL) {
+		inBufs[inbuf.cBuffers].BufferType = SECBUFFER_CHANNEL_BINDINGS;
+		inBufs[inbuf.cBuffers].pvBuffer = sec_pkg_context_bindings->Bindings;
+		inBufs[inbuf.cBuffers].cbBuffer = sec_pkg_context_bindings->BindingsLength;
+		inbuf.cBuffers++;
+	}
+
+	tokenBufferIndex = inbuf.cBuffers;
+	haveToken = TRUE;
+	inBufs[tokenBufferIndex].BufferType = SECBUFFER_TOKEN;
+	inBufs[tokenBufferIndex].pvBuffer = base64_decode(challenge, &len);
+	if (!inBufs[tokenBufferIndex].pvBuffer) {
+		return AUTH_GSS_ERROR;
+	}
+	inBufs[tokenBufferIndex].cbBuffer = len;
+	inbuf.cBuffers++;
+
+	/* Get cbMaxToken size for the output buffer */
+	PSecPkgInfo pkgInfo;
+	status = QuerySecurityPackageInfo(wide_to_utf8(state->mechoid), &pkgInfo);
+	if (status != SEC_E_OK) {
+		set_gsserror(status, "QuerySecurityPackageInfo");
+		status = AUTH_GSS_ERROR;
+		goto done;
+	}
+
+	outbuf.ulVersion = SECBUFFER_VERSION;
+	outbuf.cBuffers = 1;
+	outbuf.pBuffers = outBufs;
+	outBufs[0].pvBuffer = NULL;
+	outBufs[0].cbBuffer = pkgInfo->cbMaxToken;
+	outBufs[0].BufferType = SECBUFFER_TOKEN;
+
+	Py_BEGIN_ALLOW_THREADS
+		status = AcceptSecurityContext(/* CredHandle */
+			&state->cred,
+			/* CtxtHandle (NULL on first call) */
+			state->haveCtx ? &state->ctx : NULL,
+			/* Buff */
+			&inbuf,
+			/* Flags */
+			ASC_REQ_INTEGRITY | ASC_REQ_SEQUENCE_DETECT | ASC_REQ_REPLAY_DETECT | ASC_REQ_CONFIDENTIALITY | state->flags,
```

Comment:
PyKerberos sets this to NULL. I'm not sure what the implications of setting these flags is. It seems like the client can set them to require integrity, confidentiality, etc. without the server having them set. How did you settle on this list?

https://github.com/apple/ccs-pykerberos/blob/PyKerberos-1.3.0/src/kerberosgss.c#L725